File: src/go/problems/sum_numbers_test.go
// https://leetcode.com/problems/sum-root-to-leaf-numbers

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type SumNumbers struct{}

func TestSumNumbers(t *testing.T) {
	gen := core.TestSuite[SumNumbers]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param(&TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}).Result(25)
	}).Add(func(tc *core.TestCase) {
		tc.Param(&TreeNode{Val: 4, Left: &TreeNode{Val: 9, Left: &TreeNode{Val: 5}, Right: &TreeNode{Val: 1}}, Right: &TreeNode{Val: 0}}).Result(1026)
	}).Run(t)
}

func (SumNumbers) Solution1(root *TreeNode) int {
	stack := []*TreeNode{}
	sum := 0

	node := root
	for node != nil || len(stack) > 0 {
		prevNode := node
		if node != nil {
			if node.Left == nil && node.Right == nil {
				sum += node.Val
			}

			stack = append(stack, node)
			node = node.Left
		} else {
			prevNode = stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			node = prevNode.Right
		}

		if node != nil && prevNode != nil {
			node.Val += prevNode.Val * 10
		}
	}

	return sum
}

func (SumNumbers) Solution2(root *TreeNode) int {
	var sum func(*TreeNode, int) int
	sum = func(node *TreeNode, accum int) int {
		if node == nil {
			return 0
		}

		accum += node.Val

		if node.Left == nil && node.Right == nil {
			return accum
		}

		accum *= 10

		return sum(node.Left, accum) + sum(node.Right, accum)
	}

	return sum(root, 0)
}

File: src/python/problems/sum_numbers_test.py
# https://leetcode.com/problems/sum-root-to-leaf-numbers

from typing import Optional
from core.problem_base import *
from models.tree_node import TreeNode

class SumNumbers(ProblemBase):
    def Solution1(self, root: Optional[TreeNode]) -> int:
        stack = []
        sum = 0

        node = root
        while node or stack:
            prevNode = node
            if node:
                if not node.left and not node.right:
                    sum += node.val

                stack.append(node)
                node = node.left
            else:
                prevNode = stack.pop()
                node = prevNode.right

            if node and prevNode:
                node.val += prevNode.val * 10

        return sum
    
    def Solution2(self, root: Optional[TreeNode]) -> int:
        def GetSum(node, accum):
            if not node:
                return 0

            accum += node.val

            if not node.left and not node.right:
                return accum

            accum *= 10

            return GetSum(node.left, accum) + GetSum(node.right, accum)

        return GetSum(root, 0)

if __name__ == '__main__':
    TestGen(SumNumbers) \
        .Add(lambda tc: tc.ParamTreeNode([1,5,2,3,4,5,6,7,8,9]).Result(4875)) \
        .Add(lambda tc: tc.ParamTreeNode([4,9,0,5,1]).Result(1026)) \
        .Add(lambda tc: tc.ParamTreeNode([1,2,3]).Result(25)) \
        .Add(lambda tc: tc.ParamTreeNode([9]).Result(9)) \
        .Run()
